File: backend/src/main/kotlin/chatter/models/Ws.kt
package chatter.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
sealed interface WsCommand {
    @Serializable
    @SerialName("authenticate")
    data class Authenticate(val token: String) : WsCommand

    // enter a team by the slug
    @Serializable
    @SerialName("enterTeam")
    data class EnterTeam(val teamSlug: String) : WsCommand

    // enter a channel by the slug
    @Serializable
    @SerialName("enterChannel")
    data class EnterChannel(val channelSlug: String) : WsCommand

    // enter a team and channel by slug at the same time
    // can happen when a user enters the app by url
    @Serializable
    @SerialName("enterTeamAndChannel")
    data class EnterTeamAndChannel(
        val teamSlug: String,
        val channelSlug: String
    ) : WsCommand

    @Serializable
    @SerialName("leaveTeam")
    data object LeaveTeam : WsCommand

    @Serializable
    @SerialName("leaveChannel")
    data object LeaveChannel : WsCommand

    @Serializable
    @SerialName("sendTextMessage")
    data class SendTextMessage(val message: String) : WsCommand
}

// in the future we could send the client more notifications than just the messages
@Serializable
sealed interface WsEvent {
    // Error happened during the handling of the commands
    @Serializable
    @SerialName("wsError")
    data class Error(val message: String, val code: Int) : WsEvent


    @Serializable
    @SerialName("messageReceived")
    data class MessageReceived(
        val message: Message,
    ) : WsEvent

    @Serializable
    @SerialName("participantListChanged")
    data class ParticipantListChanged(
        val teamSlug: String
    ) : WsEvent

    @Serializable
    @SerialName("channelListChanged")
    data class ChannelListChanged(
        val teamSlug: String
    ) : WsEvent
}

File: frontend/src/models/ws.ts
import { z } from "zod";
import { message } from "./messages";

const enterTeamCommand = z.object({
  type: z.literal("enterTeam"),
  teamSlug: z.string(),
});

const enterChannelCommand = z.object({
  type: z.literal("enterChannel"),
  channelSlug: z.string(),
});

const enterTeamAndChannelCommand = z.object({
  type: z.literal("enterTeamAndChannel"),
  teamSlug: z.string(),
  channelSlug: z.string(),
});

const leaveTeamCommand = z.object({ type: z.literal("leaveTeam") });
const leaveChannelCommand = z.object({ type: z.literal("leaveChannel") });

const authenticateCommand = z.object({
  type: z.literal("authenticate"),
  token: z.string(),
});

const sendMessageCommand = z.object({
  type: z.literal("sendTextMessage"),
  message: z.string(),
});

export type WsCommand = z.infer<typeof wsCommand>;
export const wsCommand = z.discriminatedUnion("type", [
  enterTeamCommand,
  enterChannelCommand,
  enterTeamAndChannelCommand,
  leaveTeamCommand,
  leaveChannelCommand,
  authenticateCommand,
  sendMessageCommand,
]);

const errorEvent = z.object({
  type: z.literal("wsError"),
  message: z.string(),
  code: z.number(),
});

const messageRecievedEvent = z.object({
  type: z.literal("messageReceived"),
  message,
});

const channelListChangedEvent = z.object({
  type: z.literal("channelListChanged"),
  teamSlug: z.string(),
});

const participantListChangedEvent = z.object({
  type: z.literal("participantListChanged"),
  teamSlug: z.string(),
});

export type WsEvent = z.infer<typeof wsEvent>;
export const wsEvent = z.discriminatedUnion("type", [
  errorEvent,
  messageRecievedEvent,
  channelListChangedEvent,
  participantListChangedEvent,
]);
